### ISSUE TITLE
memory leak in mkl fft useage. 

### ISSUE THREAD

**Issue by ahnitz** [gwastro/pycbc#2202]
https://github.com/gwastro/pycbc/blob/master/pycbc/fft/mkl.py#L112

It should have been fixed by the above, but this seems not to be the case for mkl. Not sure what is causing the leak at this point, but it does not occur with FFTW.

This was observed in pycbc live testing.

**Comment by josh-willis** [comment]
@ahnitz Just to be sure, in the tests, you are calling the function API as at the line above, and not the class API? Do you have a test case you can point me to that reproduces the problem?

**Comment by josh-willis** [comment]
OK. Well, I'll wait for you to come up with a test case, and then try to run under some debugging tools. Since, like you, it seems to me that the obvious sources of memory leaks have been fixed, I'm not sure how to know what's causing the problem without doing that.

**Comment by ahnitz** [comment]
Reproducing case.

```
import pycbc.types
from pycbc.scheme import MKLScheme
from pycbc.fft import fft

a = pycbc.types.zeros(2**10, dtype=pycbc.types.complex64)
b = a * 1

ctx = MKLScheme()

with ctx:
    while 1:
        fft(a, b)
```

**Comment by ahnitz** [comment]
@josh-willis Did you ever have a chance to reproduce this?

**Comment by josh-willis** [comment]
> @josh-willis Did you ever have a chance to reproduce this?

@ahnitz I'll take a look this week or early next, but I have not yet tried to reproduce it.

**Comment by josh-willis** [comment]
> > @josh-willis Did you ever have a chance to reproduce this?
> 
> @ahnitz I'll take a look this week or early next, but I have not yet tried to reproduce it.

So I can reproduce it.  I'll start trying to debug it.

**Comment by josh-willis** [comment]
Still have not found the cause of this.  I have tried the following:
 1. Running under various Python memory debugging tools, which have not shown anything particularly suspicious.
 1. I have added code to explicitly delete some temporary arrays and the variable that holds the status from MKL function calls (shouldn't have been necessary, as these should have been garbage collected)
 1. Following [this guidance](https://software.intel.com/en-us/mkl-linux-developer-guide-avoiding-memory-leaks-in-intel-mkl), I have tried both setting `MKL_DISABLE_FAST_MM=1` , calling `mkl_free_buffers()`, and calling `mkl_thread_free_buffers()`. None of these has impacted the leak.
